crates/red_knot_python_semantic/src/semantic_index/expression.rs

```diff
@@ -33,23 +33,19 @@ pub(crate) enum ExpressionKind {
 #[salsa::tracked]
 pub(crate) struct Expression<'db> {
     /// The file in which the expression occurs.
-    #[id]
     pub(crate) file: File,
 
     /// The scope in which the expression occurs.
-    #[id]
     pub(crate) file_scope: FileScopeId,
 
     /// The expression node.
-    #[no_eq]
     #[return_ref]
+    #[tracked]
     pub(crate) node_ref: AstNodeRef<ast::Expr>,
```

Comment:
It's a bit unfortunate but we still need to keep the `tracked` for the `AstNodeRef` itself as @sharkdp test demonstrated it. The problem otherwise is that all definitions and expressions get new ids in the next revision, which in turn invalidates all `AttributeAssignment`, and the `attribute_assignments` query. The only way around this is if we use ids instead of actual ast nodes, ideally ids that are numbered per scope instead of globally to avoid that inserting an expression at the top invalidates the ids of all lower expressions.

Reply:
...and using IDs also requires that we allocate the AST in an arena so we have free efficient lookup from ID to node, otherwise the lookup becomes expensive.

This makes sense; should we document it in a code comment?

Reply:
Hmm, not sure where the best place is. I'll add it to `AstNodeRef` because that's the only "common" place or I'd have to document it on every tracked struct